pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -83,10 +85,43 @@ func (s *preFilterState) Clone() framework.StateData {
 // failed by this plugin schedulable.
 func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
 
+// isSchedulableAfterNodeChange is invoked whenever a node changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
```

Comment:
@alculquicondor 

I have one question https://github.com/kubernetes/kubernetes/pull/119155#discussion_r1262206772

> I would prefer to pass a logger, for performance reasons.

I remember we talked about ctx somewhere in the past but couldn't find the thread.
Why do we eliminate ctx from the parameters? In wasm project, we're working on QueueHint support and communication to the guest needs ctx.

Reply:
The hint functions are in a critical path, so they have to be as performant as possible.

Extracting things from a context (such as the logger) is sometimes an expensive operation. I would say that logging itself is an expensive operation, so any logger usage within a hint should require a high `V`.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging#performance-overhead

Reply:
logger is only one feature of context. I think we maybe should judge about the cost of the parameter separate from an assumption it will be used with a logger. I highly doubt benchmarks will show the parameter has a significant cost host-side. If we don't make choices like this with benchmarks, we risk conjecture. I've been proven wrong many times about things like this.

Reply:
It will depend on how deep in a chain of contexts the logger is. The author of KEP 3077 did extensive benchmarking already.

Reply:
I think I made a mistake giving the example of logger (I was suggested to use this example by someone, but it ended up not helping).

Concretely wasm has at least a cancel/done context use case. As it is third party code, compiled to native code, it needs to be able to know when to stop (in case endless loop or whatever). There are other use cases such as enabling experimental features or conditionally enabling profiling. What I meant by benchmarking wasn't about logging I meant just the context param existing and propagating it. I figured that host side this wouldn't be an issue. Does this help?

Reply:
in other words, can we allow context propagated like we do all the other callbacks? This is the odd one out and it doesn't seem congruent and also hurts our ability to implement when one of a dozen types of methods, some used far more often than this, rejects context.

Reply:
> it needs to be able to know when to stop (in case endless loop or whatever)

That's not something that you should do for scheduling hints.

> There are other use cases such as enabling experimental features or conditionally enabling profiling.

Those sounds like more valid use cases.

> What I meant by benchmarking wasn't about logging I meant just the context param existing and propagating it.

Propagating by itself is not costly. Obtaining any variable from the context could be expensive, depending on the chain. And this is exactly what we do when getting a logger from a context. In other words, any hint that needs a logger will be more costly, even if using a high `V` value (we can't know what's the value until we get the logger).

Can we start without a context and iterate in future versions?

Reply:
> Can we start without a context and iterate in future versions?

+1, let me follow up that.